app/src/main/java/ch/epfl/skysync/viewmodel/FinishedFlightsViewModel.kt

```diff
@@ -0,0 +1,135 @@
+  private val _selectedFlightId: MutableStateFlow<String?> = MutableStateFlow(null)
+  private val isLoading = MutableStateFlow(false)
+  private val _flightReports: MutableStateFlow<List<Report>?> = MutableStateFlow(null)
+
+  val currentFlights = _currentFlights.asStateFlow()
+  val currentUser = _currentUser.asStateFlow()
+  val selectedFlightId = _selectedFlightId.asStateFlow()
+  val flightReports = _flightReports.asStateFlow()
+
+  fun refresh() {
+    refreshUserAndFlights()
+  }
+
+  private fun refreshUser() =
+      viewModelScope.launch {
+        _currentUser.value = repository.userTable.get(userId ?: UNSET_ID, onError = { onError(it) })
+      }
+
+  private fun refreshFlights() =
+      viewModelScope.launch {
+        if (_currentUser.value is Admin) {
+          _currentFlights.value =
+              repository.flightTable
+                  .getAll(onError = { onError(it) })
+                  .filterIsInstance<FinishedFlight>()
+        } else if (_currentUser.value is Pilot || _currentUser.value is Crew) {
+          _currentFlights.value =
+              repository.userTable
+                  .retrieveAssignedFlights(
+                      repository.flightTable, userId ?: UNSET_ID, onError = { onError(it) })
+                  .filterIsInstance<FinishedFlight>()
+        }
+      }
+
+  /** Refreshes the user logged in and its finished flights */
+  fun refreshUserAndFlights() =
+      viewModelScope.launch {
+        isLoading.value = true
+        refreshUser().join()
+        refreshFlights().join()
+        isLoading.value = false
+      }
+
+  fun addFlight(flight: FinishedFlight) =
+      viewModelScope.launch {
+        repository.flightTable.add(flight, onError = { onError(it) })
+        refreshUserAndFlights()
+      }
+
+  fun deleteFlight(flightId: String) =
+      viewModelScope.launch {
+        repository.flightTable.delete(flightId, onError = { onError(it) })
+        refreshUserAndFlights()
+      }
+
+  fun getFlight(flightId: String): StateFlow<FinishedFlight?> {
+    return _currentFlights
+        .map { flights -> flights?.find { it.id == flightId } }
+        .stateIn(scope = viewModelScope, started = WhileUiSubscribed, initialValue = null)
+  }
+
+  fun addReport(report: Report) =
+      viewModelScope.launch {
+        if (_selectedFlightId.value == null) {
+          onError(Exception("No flight selected"))
+          return@launch
+        }
+        repository.reportTable.add(report, _selectedFlightId.value!!, onError = { onError(it) })
+      }
+
+  fun getAllReports(flightId: String) =
+      viewModelScope.launch {
+        _flightReports.value =
+            repository.reportTable.retrieveReports(flightId, onError = { onError(it) })
+        Log.d("FinishedFlightsViewModel", "Reports retrieved : ${_flightReports.value}")
+      }
+
+  private fun selectFlight(id: String) = viewModelScope.launch { _selectedFlightId.value = id }
+
+  fun refreshAndSelectFlight(id: String) =
```

Comment:
This function is never used (except in the tests), and you need to call this function to set `_selectedFlightId`, so `_selectedFlightId` is never set, so `addReport` will never work. Maybe you could drop the `_selectedFlightId` attribute altogether and pass a `flightId` argument to `addReport`?

Reply:
I have mixed up to ideas. I'll keep your suggestion as it will make the viewModel simpler